iocore/utils/SharedExtendible.h

```diff
@@ -0,0 +1,447 @@
+ *
+ * The size of this structure is actually zero, so it will not change the size of your derived class.
+ * But new and delete are overriden to use allocate enough bytes of the derived type + added fields.
+ * All bool's are packed to save space using the *Bit methods.
+ * This API is focused on thread safe data types that allow minimally blocked reading.
+ * This is templated so static variables are instanced per Derived type. B/c we need to have different
+ * size field sets.
+ *
+ * @tparam Derived_t - the class that you want to extend at runtime.
+ *
+ * @see test_SharedExtendible.cc for examples
+ *
+ */
+template <typename Derived_t> struct SharedExtendible {
+  using SharedExtendible_t = SharedExtendible<Derived_t>;
+
+  /////////////////////////////////////////////////////////////////////
+  /// strongly type the FieldId to avoid user error and branching logic
+  template <SharedAccessEnum FieldAccess_e, typename Field_t> class FieldId
+  {
+  private:
+    SharedOffset_t const *offset_ptr;
+
+  public:
+    using FieldId_t = FieldId<FieldAccess_e, Field_t>;
+
+    FieldId() { offset_ptr = nullptr; }
+
+    static FieldId_t
+    find(std::string const &field_name)
+    {
+      auto field_iter = schema.fields.find(field_name);
+      ink_release_assert(field_iter != schema.fields.end());
+      FieldSchema &fs = field_iter->second;
+      ink_release_assert(fs.access == FieldAccess_e);
+      return FieldId_t(fs.offset);
+    }
+
+    // only allow offest to be used by schema code
+    friend SharedExtendible_t;
+
+  private:
+    FieldId(SharedOffset_t const &offset) { offset_ptr = &offset; }
+    SharedOffset_t
+    offset() const
+    {
+      return *offset_ptr;
+    }
+  };
+  using BitFieldId = FieldId<BIT, bool>;
+
+  /////////////////////////////////////////////////////////////////////
+  /// defines a runtime "member variable", element of the blob
+  struct FieldSchema {
+    using Func_t = void(void *);
+
+    SharedAccessEnum access; ///< which API is used to access the data
+    SharedOffset_t size;     ///< size of field
+    SharedOffset_t offset;   ///< offset of field from 'this'
+    Func_t *construct_fn;    ///< the data type's constructor
+    Func_t *destruct_fn;     ///< the data type's destructor
+  };
+
+  /////////////////////////////////////////////////////////////////////
+  /// manages the fields as data structures
+  class Schema
+  {
+    friend SharedExtendible_t;
+
+  private:
+    std::unordered_map<std::string, FieldSchema> fields; ///< defined elements of the blob by name
+    uint32_t bit_offset;                                 ///< offset to first bit
+    size_t alloc_size;                                   ///< bytes to allocate
+    size_t alloc_align;                                  ///< alignment for each allocation
+    std::atomic_uint instance_count;                     ///< the number of class instances in use.
```

Comment:
Class instances of what, exactly? Not this class or this should be a `static`.

Reply:
`///< the number of SharedExtendible<Derived> instances in use.`